src/theory/bv/theory_bv_solver.h

```diff
@@ -0,0 +1,126 @@
+/*********************                                                        */
+/*! \file theory_bv_solver.h
+ ** \verbatim
+ ** Top contributors (to current version):
+ **   Mathias Preiner
+ ** This file is part of the CVC4 project.
+ ** Copyright (c) 2009-2020 by the authors listed in the file AUTHORS
+ ** in the top-level source directory) and their institutional affiliations.
+ ** All rights reserved.  See the file COPYING in the top-level source
+ ** directory for licensing information.\endverbatim
+ **
+ ** \brief Bit-vector solver interface.
+ **
+ ** Describes the interface for the internal bit-vector solver of TheoryBV.
+ **/
+
+#include "cvc4_private.h"
+
+#ifndef CVC4__THEORY__BV__THEORY_BV_SOLVER_H
+#define CVC4__THEORY__BV__THEORY_BV_SOLVER_H
+
+#include "theory/bv/theory_bv.h"
+#include "theory/theory.h"
+
+namespace CVC4 {
+namespace theory {
+namespace bv {
+
+class TheoryBVSolver
+{
+ public:
+  TheoryBVSolver(TheoryBV& bv)
+      : d_bv(bv), d_state(bv.d_state), d_inferManager(bv.d_infer){};
+
+  virtual ~TheoryBVSolver(){};
+
+  /**
+   * Returns true if we need an equality engine. If so, we initialize the
+   * information regarding how it should be setup. For details, see the
+   * documentation in Theory::needsEqualityEngine.
+   */
+  virtual bool needsEqualityEngine(EeSetupInfo& esi) { return false; }
+
+  virtual void finishInit(){};
+
+  virtual void preRegisterTerm(TNode n) = 0;
+
+  /**
+   * Forwarded from TheoryBV::preCheck().
+   */
+  virtual bool preCheck(Theory::Effort level = Theory::Effort::EFFORT_FULL)
+  {
+    return false;
+  }
+  /**
+   * Forwarded from TheoryBV::postCheck().
+   */
+  virtual void postCheck(Theory::Effort level = Theory::Effort::EFFORT_FULL){};
+  /**
+   * Forwarded from TheoryBV:preNotifyFact().
+   */
+  virtual bool preNotifyFact(
+      TNode atom, bool pol, TNode fact, bool isPrereg, bool isInternal)
+  {
+    return false;
+  }
+  /**
+   * Forwarded from TheoryBV::notifyFact().
+   */
+  virtual void notifyFact(TNode atom, bool pol, TNode fact, bool isInternal) {}
+
+  virtual bool needsCheckLastEffort() = 0;
+
+  virtual void propagate(Theory::Effort e){};
+
+  virtual TrustNode explain(TNode n)
+  {
+    Unimplemented()
+        << "TheoryBVSolver propagated a node but doesn't implement the "
+           "TheoryBVSolver::explain() interface!";
+    return TrustNode::null();
+  };
+
+  virtual bool collectModelInfo(TheoryModel* m) = 0;
```

Comment:
Notice that Theory is not recommended to implement collectModelInfo anymore. This method will be replaced by "collectModelValues" (see https://github.com/ajreynol/CVC4/blob/master/src/theory/theory.h#L699), which has a simpler contract.  However, this will require some changes to the old lazy BV solver to use this interface, which I have on my branch.

Thus, for now, I am fine with having this virtual method to make this PR go through, but I'd recommend marking this as temporary with a comment.  I can replace this method with the new one (collectModelValues) on a subsequent PR.